### PR TITLE
[ENH] Bollinger Bands

### LINKED ISSUE #6320: [ENH] Bollinger bands
I noticed `sktime` has some tools for smoothing, but not the popular Bollinger bands used in trading.

While one can be of different opinion on the estimator, I think it is popular enough to have it in the "smoothing" section.

It should be a good first issue, given the simple nature of the estimator it should be straightforward to implement from scratch, using `pandas` or `numpy` directly.
Some thoughts:

* the `q` of the moving average should be a parameter
* the `remember_data` tag should be true, since stream use will require to remember previous data for computing the bands. So, transform should try to retrieve remembered data, and subset it to the range required to comput bands for indices in the input.

### PR DESCRIPTION
**Pull request by ishanpai**
#### Reference Issues/PRs
closes #6320.

#### What does this implement/fix? Explain your changes.
Implements the Bollinger Bands for a time series

#### Does your contribution introduce a new dependency? If yes, which one?

No

#### What should a reviewer concentrate their feedback on?

- Code structure
- Code Location
- Do we need an "inv_transform" for this? When would we use that?

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [x] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [x] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
Well, that was extremely quick! Nice!

Some questions/requests:

* I think the typing should be `transform-input` and `transofrm-output` being `"Series"`
* does the `pd.concat` with `keys` not result in a column `pd.MultiIndex`, from which you may want to drop one level?
* could you add a second test parameter set? Return of `get_test_params` should be list of dict.
* location is fine imo; transformer should be added to `docs/api_reference/transformations.rst`, I would put it in the smoothing section.

**Comment by ishanpai** [comment]
> I think the typing should be transform-input and transofrm-output being "Series"

I agree that `transform-input` should be `Series` and have changed that. `transform-output `currently returns 3 columns, one with the moving average, one with the lower Bollinger bound and one with the upper Bollinger bound. Should this still return a `Series`? Technically it is a `DataFrame` so should it be `Panel`?


> does the pd.concat with keys not result in a column pd.MultiIndex, from which you may want to drop one level?

Since the name of all of the series we are concatting is the same, it results in a `DataFrame` with the column names passed in `keys`, without needing to drop one level.

> location is fine imo; transformer should be added to docs/api_reference/transformations.rst, I would put it in the smoothing section.

Maybe I am being silly, but I cannot find the "smoothing" section. Is this the "Filtering and denoising" section? I have currently added it to the "Window-based series transforms" section.

**Comment by ishanpai** [comment]
It is also not clear to me how the `remember_data` tag and `memory` work. I copied the `memory` from the `Differencer`, but I am not sure if it is necessary here if I am not implementing the inverse transformation.

File: sktime/transformations/series/bollinger.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        "fit_is_empty": False,
+        "transform-returns-same-time-index": True,
+        "univariate-only": False,
+        "capability:inverse_transform": False,
+        "remember_data": True,
+    }
+
+    def __init__(self, window, k=1, memory="all"):
+        assert window > 1, f"window must be greater than 1, passed {window}"
+        assert k > 0, f"k must be positive, passed {k}"
```
Comment (VascoSch92) [inline sktime/transformations/series/bollinger.py:66]: this are `ValueError`, but nothing against using `assert` :-)

File: sktime/transformations/series/bollinger.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+            Parameters to create testing instances of the class
+            Each dict are parameters to construct an "interesting" test instance, i.e.,
+            ``MyClass(**params)`` or ``MyClass(**params[i])`` creates a valid test
+            instance.
+            ``create_test_instance`` uses the first (or only) dictionary in ``params``
+        """
+        return [{"window": 12, "k": 1}, {"window": 2, "k": 1.2}]
+
+
+def _bollinger_transform(df, window, k):
```
[the remark this answers is not in the document]
Comment (fkiraly) [reply sktime/transformations/series/bollinger.py]: it's not necessary to do that, but sometimes cleaner.

File: sktime/transformations/series/bollinger.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        Returns
+        -------
+        Xt : pd.Series or pd.DataFrame, same type as X
+            transformed version of X
+        """
+        X_orig_index = X.index
+
+        X = update_data(X=self._X, X_new=X)
+
+        X = self._check_freq(X)
```
Comment (fkiraly) [inline sktime/transformations/series/bollinger.py:134]: could you explain why we need this?

Comment (ishanpai) [reply sktime/transformations/series/bollinger.py]: Don't think it is, I have deleted this now.

**Comment by fkiraly** [comment]
Failure is unrelated, known sporadic: https://github.com/sktime/sktime/issues/6406

File: docs/source/api_reference/transformations.rst [hunk 4]
```diff
@@ -392,6 +392,14 @@ Filtering and denoising
 
     ThetaLinesTransformer
 
+.. currentmodule:: sktime.transformations.series.bollinger
+
+.. autosummary::
+    :toctree: auto_generated/
+    :template: class.rst
+
+    Bolllinger
```
Comment (fkiraly) [inline docs/source/api_reference/transformations.rst:401]: typo - 3 L

Comment (ishanpai) [reply docs/source/api_reference/transformations.rst]: Good spot, thanks

**Review by fkiraly — APPROVED** [review]
Looks great now!

**Pull request merged**